### ISSUE TITLE
An input component for Google Custom Search Business Edition [CARROT-107]

### ISSUE THREAD

**Issue by dweiss** [carrot2/carrot2#313]
It looks like Google started to provide REST/XML interface for their index: http://www.google.com/enterprise/csbe/
At some point it may make sense to implement an input component for this API.

---
Issue: CARROT-107 (migrated from JIRA), created by Stanisław Osiński (@stanislawosinski), updated Mar 20 2012

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

The description of their XML API is here:
http://www.google.com/coop/docs/cse/resultsxml.html

Note that "The default num value is 10, and the maximum value is 20." so at least 5 requests would be required to get 100 results.

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

True. In fact, XML input component could do most of the job – we'd need to add support for some attribute along the lines of "startAt" and make more requests if the number of results returned per request is smaller than the number of results requested.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

Not necessarily true. We had and used such a feature with Yahoo! – it turned out to be more problematic than helpful. Yahoo! would return, say, 20 results for the requested 100 occasionally even if the indicated total was a lot more (in the order of thousands). Multiple follow-up requests wouldn't fetch anything. 

I think it's partially due to heuristics working on the IR backend and I think – depending on which cluster you hit you can get different sets of results and different counts. For this reason I tend to give a -1 for the feature you suggest. We should rely on user-provided information about the maximum number of results returned per request rather than compensate for this automatically (as it leads to problems).

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

Ok – I wasn't precise. Just as Yahoo! has a per-request limit of 100 and it does not prevent us from fetching 500 (user specified) results from it (by sending 5 sequential requests), we could add a similar option to XML input. Open Search input, which is quite like XML input, already has that, ans XML input is for the time being limited to whatever number of results the provider happens to allow in one request.

To sum up, I didn't suggest to make more than ceil(user-requested-results / max-results-per-request) requests :)
